src/v/archival/archiver_service.cc

```diff
@@ -0,0 +1,999 @@
+            // running the archiver is that the broker may run out of disk
+            // space. Because the data is not uploaded to the cloud storage the
+            // max_collectible_offset can't be propagated forward. Because of
+            // that the local data can't be evicted.
+            vlog(
+              fsm._ctxlog.error,
+              "ALERT! ntp_archiver failed to start {}",
+              f.error);
+            prev.clear();
+            if (managed.trigger) {
+                managed.trigger->set_exception(f.error);
+            }
+        }
+    };
+
+    struct active_to_stopping {
+        template<class Event>
+        void operator()(
+          const Event&,
+          state_machine_t& fsm,
+          st_active& active,
+          st_stopping_async&) {
+            vlog(fsm._ctxlog.info, "stopping ntp_archiver");
+            ssx::background = fsm.bg_stop_and_dispose_archiver(active.archiver)
+                                .then_wrapped([&fsm](ss::future<> fut) mutable {
+                                    if (fut.failed()) {
+                                        vlog(
+                                          fsm._ctxlog.error,
+                                          "ntp_archiver shutdown failed {}",
+                                          fut.get_exception());
+                                        fsm.process_event(ev_archiver_failure{
+                                          .error = fut.get_exception()});
+                                    } else {
+                                        fsm.process_event(
+                                          ev_archiver_stopped{});
+                                    }
+                                })
+                                .finally([arch = active.archiver] {});
+            active.clear();
+        }
+    };
+
+    struct stopping_to_managed {
+        void operator()(
+          const ev_archiver_stopped&,
+          state_machine_t& fsm,
+          st_stopping_async& prev,
+          st_managed& managed) {
+            vlog(fsm._ctxlog.info, "ntp_archiver stopped");
+            prev.clear();
+            if (managed.trigger) {
+                managed.trigger->set_value();
+            }
+        }
+        void operator()(
+          const ev_archiver_failure& f,
+          state_machine_t& fsm,
+          st_stopping_async& prev,
+          st_managed& managed) {
+            vlog(
+              fsm._ctxlog.error, "ntp_archiver failed to shutdown {}", f.error);
+            prev.clear();
+            if (managed.trigger) {
+                managed.trigger->set_exception(f.error);
+            }
+        }
+    };
+
+    template<class Src, class Event, class Target, class Action>
+    using row = boost::msm::front::Row<Src, Event, Target, Action>;
+    using none = boost::msm::front::none;
+    using defer = boost::msm::front::Defer;
+    // clang-format off
+    using transition_table = boost::mpl::vector<
```

Comment:
I suppose this has a compile time assertion to be valid and that there are no duplicate rows for the same source state + event? Maybe add a tiny comment about what this is and how it used by MSM for newbies like me.

Reply:
I see the comment. Have you measured the actual compile time cost? It is in a CC file, how expensive can it be? 😄

Alternatively, can we use this table with the alternative(mpl_graph_fsm_check) backend in a unit test? We'd have it compiled in CI and when you want to run a specific unit test but in most of the cases (dev-build-run cycle) we wouldn't need to recompile it.